sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -485,6 +486,46 @@ object RemoveRedundantAliases extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = removeRedundantAliases(plan, AttributeSet.empty)
 }
 
+/**
+ * Remove redundant aggregates from a query plan. A redundant aggregate is an aggregate whose
+ * only goal is to keep distinct values, while its parent aggregate would ignore duplicate values.
+ */
+object RemoveRedundantAggregates extends Rule[LogicalPlan] with AliasHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case upper @ Aggregate(_, _, lower: Aggregate) if lowerIsRedundant(upper, lower) =>
+      val aliasMap = getAliasMap(lower)
+
+      val newAggregate = upper.copy(
+        child = lower.child,
+        groupingExpressions = upper.groupingExpressions.map(replaceAlias(_, aliasMap)),
+        aggregateExpressions = upper.aggregateExpressions.map(
+          replaceAliasButKeepName(_, aliasMap))
+      )
+
+      // We might have introduces non-deterministic grouping expression
+      if (newAggregate.groupingExpressions.exists(!_.deterministic)) {
+        PullOutNondeterministic.applyLocally.applyOrElse(newAggregate, identity[LogicalPlan])
+      } else {
+        newAggregate
+      }
+  }
+
+  private def lowerIsRedundant(upper: Aggregate, lower: Aggregate): Boolean = {
+    val upperReferencesOnlyGrouping = upper.references.subsetOf(AttributeSet(
+      lower.aggregateExpressions.filter(!isAggregate(_)).map(_.toAttribute)))
+
+    val upperHasNoAggregateExpressions = upper.aggregateExpressions
+      .forall(_.find(isAggregate).isEmpty)
+
+    upperReferencesOnlyGrouping && upperHasNoAggregateExpressions
```

Comment:
nit: to avoid unnecessary object creation (e.g,. `AttributeSet`) how about reorganizing it like this?
```
  private def lowerIsRedundant(upper: Aggregate, lower: Aggregate): Boolean = {
    val upperHasNoAggregateExpressions = upper.aggregateExpressions
      .forall(_.find(isAggregate).isEmpty)
    lazy val upperRefsHaveGroupingOnly = upper.references.subsetOf(AttributeSet(
      lower.aggregateExpressions.filter(!isAggregate(_)).map(_.toAttribute)))
    upperHasNoAggregateExpressions && upperRefsHaveGroupingOnly 
  }
```